### PR TITLE
Migrates listbox class names to .select;

### LINKED ISSUE #222: Select: retrofit into DS4
Essentially this means renaming `Listbox` to `Select` in DS4.

### PR DESCRIPTION
**Pull request by seangates**
## Description
- Migrates all `.listbox` classnames to `.select`
- ~Updates for markup changes in the ebayui-core `ebay-select` component.~ (see bullet 5 below)
- Normalizes common selectors and rules into a `/base/select.less`
- Handles additional modifiers for DS6 select
- Supports both background and foreground SVG (removes the DS6 `::after` pseudo-element), requiring a slight markup change

## Context
Needed for DS4 being supported in ebayui-core.

## References
Fixes #222 

## Screenshots
image
image
image

### CODE REVIEW

**Comment by ianmcburnie** [comment]
I take it this a partial fix for #222 and then you will be filing more PRs for the rest of the changes (docs, markup, source folders, test pages, etc)?

File: docs/_includes/ds4/listbox.html [hunk 1]
```diff
@@ -6,145 +6,127 @@ <h2>Listbox</h2>
...
-                    </svg>
-                </span>
             </span>
         </div>
     </div>
 
     {% highlight html %}
-<span class="listbox">
-    <select class="listbox__control listbox__control--borderless">
+<span class="select select--borderless\">
```
Comment (ianmcburnie) [inline docs/_includes/ds4/listbox.html]: Typo. Stray `\` in class.

**Comment by seangates** [comment]
@ianmcburnie Not sure what you mean. This has source, docs and markup changes.

File: docs/_includes/ds4/listbox.html (hunk 1 above)
Comment (seangates) [reply docs/_includes/ds4/listbox.html]: Will fix.

**Comment by ianmcburnie** [comment]
But it still says "Listbox"everywhere.

**Comment by seangates** [comment]
Ah. Gotcha. The nomenclature, not the class names. I can fix that in this PR, sure.

**Comment by ianmcburnie** [comment]
Also, it's not following the pattern of `base`, `ds4` and `ds6` in the `src\select` folder like I started with checkbox and radio.

And the markup should support background SVG and foreground SVG. I added a comment in the issue about this.

You'll also need to update the ds4 test page with the new classnames and nomenclature.

**Comment by ianmcburnie** [comment]
This may have got broken along the way (or we may have missed it in a prior PR).

screen shot 2018-07-14 at 9 52 56 am

**Comment by seangates** [comment]
👍 Workin' on all of it.

**Comment by ianmcburnie** [comment]
Feel free to break it out into separate PRs if you want to keep the scope of this one to just changing `.listbox` to `.select` as per title.

Also I would prefer if we leave the old `.listbox` around (and listbox.browser.json) for awhile until we are sure we want to delete it (and remove backwards compatibility for folks still using it).

**Comment by seangates** [comment]
Okay, I've made some updates this evening (in prep for tomorrow). Let me know your thoughts, @ianmcburnie.

A couple of things to note from your previous comments:
- The listbox CSS is still intact, but no longer used for the examples
- Background SVG with `::after` pseudo-element causes problems for both disabled state and for custom icons. We would need to modify the `ebay-select` source to make the disabled icons work properly
- Everything which can be shared is now in a `/select/base/select.less` file and shared with their respective `/select/ds{#}/select.less` file.
- Added the disabled state and corresponding example for DS6, but it isn't in the DSL-provided pattern. It was the only state I added, preferring not to add the `aria-invalid="true"` nor the `[readonly]` states since there is not a great use case for those at the moment. Please let me know if you feel they should be included in DS6.

Looking forward to comments.

'Night! 💤

**Comment by ianmcburnie** [comment]
Does this support foreground SVG for the icon? I don't see it. It seems like you have normalized on the DS6 markup for select, which is fine for this PR, but I feel like the DS4 markup is better as it supports background and foreground SVG using a real element.

**Comment by seangates** [comment]
Current DS4 supports a foreground SVG. But, the current `ebay-select` markup does not support it. That was my question earlier: should I also modify the `ebay-select` to support a foreground icon.

**Comment by ianmcburnie** [comment]
Yep, ebay-select will need updating for coreui v1.0.0 (so, no hurry just yet) to use the new markup.

**Comment by seangates** [comment]
👍 Will make this change and update the PR.

**Comment by seangates** [comment]
@ianmcburnie Made discussed changes. Please review once more. I think this is really close to final, so let me know if you find anything.

**Comment by ianmcburnie** [comment]
Loving this:

screen shot 2018-07-16 at 8 48 38 pm

Been wanting to add some RTL examples for a while now. Nice! Eventually we will go through all components and do this. Most of them have been setup for it.

**Comment by ianmcburnie** [comment]
Let's try and normalize the docs and examples if we can (like I did with checkbox and radio). Or leave it for a follow up PR if you prefer.

File: dist/listbox/ds4/listbox.css [hunk 2]
```diff
@@ -1,14 +1,36 @@
-.listbox {
+.select {
+  height: 40px;
   position: relative;
 }
-span.listbox {
+.select::after {
```
Comment (ianmcburnie) [inline dist/listbox/ds4/listbox.css]: Why did you change the existing DS4 implementation of listbox to use `::after`? We should just leave the existing code as is (using span or svg element) so there are no breaking changes until we decide to officially remove it.

File: src/less/listbox/ds4/listbox-base.less [hunk 3]
```diff
@@ -1,145 +0,0 @@
-// listbox text should NOT scale with root font-size, so all font-size defined with px
```
Comment (ianmcburnie) [inline src/less/listbox/ds4/listbox-base.less]: I don't think we had officially decided to delete this yet. Leaving it in for backwards compatibility doesn't hurt.

File: src/less/select/ds4/select.less [hunk 4]
```diff
@@ -0,0 +1,74 @@
+@import "../../less/ds4/variables.less";
+@import "../../less/ds4/mixins.less";
+
+@import "../base/select.less";
+
+.select__icon:empty {
```
Comment (ianmcburnie) [inline src/less/select/ds4/select.less]: Hmm, we're using `span` element selector everywhere else rather than `:empty` for the background SVG. Reason being that we keep the specificity the same when we need to add any `svg` element rules. Not sure if `:not(:empty)` would have the same specificity but I think the element selector is more readable. Maybe we should stick with consistency for now?

File: docs/static/ds6/skin.css [hunk 5]
```diff
@@ -2114,57 +2114,115 @@ span.inline-notice {
...
-          appearance: none;
   background-color: #fff;
-  border: 1px solid #c7c7c7;
+  border-color: #c7c7c7;
   border-radius: 0;
   font-family: "Market Sans", Arial, sans-serif;
-  font-size: 16px;
-  height: 40px;
-  margin: 0;
+  font-size: 1rem;
```
Comment (ianmcburnie) [inline docs/static/ds6/skin.css]: I'm a bit nervous about the property changes, as they got lost in the diff and I can't compare them all. I would try and minimize refactoring of properties and save them for a separate PR where they are easier to diff.  I do know we want to keep select at 16px for now to match button and textbox (which are also hardcoded).

Make sure that you update and use the existing test page. There should be no regressions (right now it's in a broken state):

screen shot 2018-07-16 at 9 06 30 pm

Compare with:

screen shot 2018-07-16 at 9 08 10 pm

File: dist/listbox/ds4/listbox.css (hunk 2 above)
Comment (ianmcburnie) [reply dist/listbox/ds4/listbox.css]: Oh. I see the listbox.css file is now getting all the new `.select` classname rules. But we should leave the existing listbox.css file in place for backwards compatibility. Let's point the new `select` module to a `select.css` only.

File: dist/listbox/ds6/listbox.css [hunk 6]
```diff
@@ -1,62 +0,0 @@
-.listbox {
```
Comment (ianmcburnie) [inline dist/listbox/ds6/listbox.css]: Yep. This one can be deleted as we never really had a DS6 listbox.

File: src/less/listbox/ds4/listbox-base.less (hunk 3 above)
Comment (seangates) [reply src/less/listbox/ds4/listbox-base.less]: Okay, I'll put it back.

File: src/less/select/ds4/select.less (hunk 4 above)
Comment (seangates) [reply src/less/select/ds4/select.less]: Can you show me an example of what you mean?

But, yes, I want to make sure it's consistent.

File: dist/listbox/ds4/listbox.css (hunk 2 above)
Comment (seangates) [reply dist/listbox/ds4/listbox.css]: Yup. Somehow this got generated and then I deleted the `listbox.less` which built it. So ... it's kinda a stray until I reinstate `listbox.less`.

File: dist/listbox/ds6/listbox.css (hunk 6 above)
Comment (seangates) [reply dist/listbox/ds6/listbox.css]: 👍

File: docs/static/ds6/skin.css (hunk 5 above)
Comment (seangates) [reply docs/static/ds6/skin.css]: I've fixed the test pages.

Also, I'm looking at why that property changed. It was not intentional as I was aiming to keep all the properties the same. In the shuffle this one might have been updated.

Comment (seangates) [reply docs/static/ds6/skin.css]: It got changed to this: `font-size: @font-size-16;`. I've changed it back to a hardcoded `16px`.

Comment (ianmcburnie) [reply docs/static/ds6/skin.css]: Ok great. We should be good if it was your aim to keep them the same.

File: src/less/select/ds4/select.less (hunk 4 above)
Comment (ianmcburnie) [reply src/less/select/ds4/select.less]:
```css
// base
.select__icon {
  ..
}
// background specific
span.select__icon {
  ...
}
// foreground specific
svg.select__icon {
  ...
}
```

Rather than

```css
// base
.select__icon {
  ..
}
// background specific
.select__icon:empty {
  ...
}
// foreground specific
svg.select__icon {
  ...
}
```

Just to be consistent with elsewhere.

**Comment by ianmcburnie** [comment]
Old listbox page is broken:

screen shot 2018-07-17 at 11 43 47 am

We should keep that working until we decide to permanently remove the DS4 listbox.

**Comment by ianmcburnie** [comment]
You should probably duplicate that current listbox test page and repurpose it for a ds4 select test page (we'll start creating the ds6 tests soon).

File: dist/listbox/ds4/listbox.css [hunk 7]
```diff
@@ -16,6 +16,7 @@ select.listbox__control {
   font-family: "Helvetica Neue", Helvetica, Arial, Roboto, sans-serif;
   font-size: inherit;
   height: 40px;
+  line-height: 1.15;
```
Comment (ianmcburnie) [inline dist/listbox/ds4/listbox.css]: I removed the line-height in this PR remember:

https://github.com/eBay/skin/pull/252/files#diff-a919e64bc0f0127334b9fabddcb84118L19

Any reason you are putting it back?

Comment (seangates) [reply dist/listbox/ds4/listbox.css]: Unintended. Will fix.

Comment (seangates) [reply dist/listbox/ds4/listbox.css]: I pulled that file back from `master`, not the 6.0.0. Whoops. ;-)

**Comment by seangates** [comment]
@ianmcburnie Okay, all PR comments should be addressed now. Let me know if you have any questions or see anything.

**Comment by ianmcburnie** [comment]
Everything looking great.... just one thing I see in Safari:

screen shot 2018-07-18 at 7 54 46 pm


screen shot 2018-07-18 at 7 57 54 pm


I don't see that in DS6:

screen shot 2018-07-18 at 7 57 37 pm

**Comment by seangates** [comment]
@ianmcburnie It isn’t in there currently for DS6. Should I add it?

**Comment by ianmcburnie** [comment]
I mean I don't see that issue in DS6.

**Comment by seangates** [comment]
@ianmcburnie Oh, in DS4 it was intentional. And DS6 doesn’t specify.

**Comment by seangates** [comment]
Ergo, I left hem as they were from before the normalization.

**Comment by ianmcburnie** [comment]
Merging this now to unblock us. We can fix the last remaining issue before launch.

**Comment by ianmcburnie** [comment]
I'll create a quick ticket for it.
